src/chromium/crBrowser.ts

```diff
@@ -50,24 +50,35 @@ export class CRBrowser extends BrowserBase {
     const connection = new CRConnection(SlowMoTransport.wrap(transport, slowMo));
     const browser = new CRBrowser(connection);
     const session = connection.rootSession;
-    const promises = [
-      session.send('Target.setDiscoverTargets', { discover: true }),
-      session.send('Target.setAutoAttach', { autoAttach: true, waitForDebuggerOnStart: true, flatten: true }),
-      session.send('Target.setDiscoverTargets', { discover: false }),
-    ];
+    if (!isPersistent) {
+      await session.send('Target.setAutoAttach', { autoAttach: true, waitForDebuggerOnStart: true, flatten: true });
+      return browser;
+    }
+
     const existingPageAttachPromises: Promise<any>[] = [];
-    if (isPersistent) {
-      // First page and background pages in the persistent context are created automatically
-      // and may be initialized before we enable auto-attach.
-      function attachToExistingPage({targetInfo}: Protocol.Target.targetCreatedPayload) {
-        if (targetInfo.type !== 'page' && targetInfo.type !== 'background_page')
-          return;
-        existingPageAttachPromises.push(session.send('Target.attachToTarget', {targetId: targetInfo.targetId, flatten: true}));
-      }
-      session.on('Target.targetCreated', attachToExistingPage);
-      Promise.all(promises).then(() => session.off('Target.targetCreated', attachToExistingPage)).catch(debugError);
+    // First page and background pages in the persistent context are created automatically
+    // and may be initialized before we enable auto-attach.
+    function attachToExistingPage({targetInfo}: Protocol.Target.targetCreatedPayload) {
+      if (targetInfo.type !== 'page' && targetInfo.type !== 'background_page')
+        return;
+      // TODO: should we handle the error during 'Target.attachToTarget'? Can the target disappear?
+      existingPageAttachPromises.push(session.send('Target.attachToTarget', {targetId: targetInfo.targetId, flatten: true}));
     }
-    await Promise.all(promises);
+    session.on('Target.targetCreated', attachToExistingPage);
+
+    const startDiscover = session.send('Target.setDiscoverTargets', { discover: true });
+    const autoAttachAndStopDiscover = session.send('Target.setAutoAttach', { autoAttach: true, waitForDebuggerOnStart: true, flatten: true }).then(() => {
+      // All targets collected before setAutoAttach response will not be auto-attached, the rest will be.
```

Comment:
Can you help me understand how it can happen? Does it imply that between handling `Target.setAutoAttach` and `Target.setDiscoverTargets` (sent synchronously one after another ) the browser may create new targets?

Reply:
Yes, we send them synchronously, but they arrive to the browser with arbitrary delay.